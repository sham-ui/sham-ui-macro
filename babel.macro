/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const inject = require( './macros/inject' );
const options = require( './macros/options' );

module.exports = createMacro( macro );

function macro( { references } ) {
    if ( references.inject ) {
        references.inject.forEach( referencePath => {
            inject( referencePath.findParent( x => x.isDecorator() ) );
        } );
    }

    if ( references.options ) {
        references.options.forEach( referencePath => {
            options( referencePath.findParent( x => x.isDecorator() ) );
        } );
    }
}
