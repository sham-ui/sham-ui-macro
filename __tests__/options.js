/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: 'method',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options text() {
                    return this.ID;
                }
            }
        `
    }, {
        title: 'ClassProperty',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options text = 'default text';
            }
        `
    }, {
        title: 'ClassProperty (array)',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options errors = [ 'errors must be overriden' ];
            }
        `
    }, {
        title: 'getter',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options get startDate() {
                    return new Date();
                }
            }
        `
    }, {
        title: 'getter & setter',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options get text() {
                    return this._text;
                }
                @options set text( value ) {
                    this._text = value;
                }
                
                @options get date() {
                    return this._date;
                }
                @options set date( value ) {
                    this._date = value;
                }
            }
        `
    }, {
        title: 'already has configureOptions',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options get startDate() {
                    return new Date();
                }
                
                configureOptions() {
                    super.configureOptions( ...arguments ); 
                }
            }
        `
    }, {
        title: 'add to existed method',
        code: `
            import { configureOptions } from 'sham-ui';
            import { options } from '../babel.macro';
            
            class FooComponent extends Template {
                @options errors = [ 'errors must be overriden' ];
            
                configureOptions() {
                    super.configureOptions(...arguments);
                    configureOptions(FooComponent.prototype, this, {
                        startDate: {
                            get() {
                                return new Date();
                            }
            
                        }
                    });
                }
            }
        `
    }, {
        title: 'complex (from doc)',
        code: `
            import { options } from '../babel.macro';

            class FooComponent extends Template {
                @options submitText = 'Submit';
                @options loaded = false;
                @options errors = [];
                @options get startDate() {
                    return new Date();
                }
                @options data = {};
                @options onLoad() {};
                
                @options get text() {
                    return this._text;
                }
                @options set text( value ) {
                    this._text = value;
                }
            }
        `
    } ]
} );
