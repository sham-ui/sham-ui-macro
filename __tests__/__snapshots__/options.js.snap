// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros ClassProperty (array): ClassProperty (array) 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options errors = [ 'errors must be overriden' ];
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            errors: ['errors must be overriden']
        });
    }

}

`;

exports[`macros ClassProperty: ClassProperty 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options text = 'default text';
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            text: 'default text'
        });
    }

}

`;

exports[`macros add to existed method: add to existed method 1`] = `

import { configureOptions } from 'sham-ui';
import { options } from '../babel.macro';

class FooComponent extends Template {
    @options errors = [ 'errors must be overriden' ];

    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            startDate: {
                get() {
                    return new Date();
                }

            }
        });
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {

    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            startDate: {
                get() {
                    return new Date();
                }

            },
            errors: ['errors must be overriden']
        });
    }
}

`;

exports[`macros already has configureOptions: already has configureOptions 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options get startDate() {
        return new Date();
    }
    
    configureOptions() {
        super.configureOptions( ...arguments ); 
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {

    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            startDate: {
                get() {
                    return new Date();
                }

            }
        });
    }
}

`;

exports[`macros complex (from doc): complex (from doc) 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options submitText = 'Submit';
    @options loaded = false;
    @options errors = [];
    @options get startDate() {
        return new Date();
    }
    @options data = {};
    @options onLoad() {};
    
    @options get text() {
        return this._text;
    }
    @options set text( value ) {
        this._text = value;
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            submitText: 'Submit',
            loaded: false,
            errors: [],
            startDate: {
                get() {
                    return new Date();
                }

            },
            data: {},

            onLoad() {},

            text: {
                get() {
                    return this._text;
                },

                set(value) {
                    this._text = value;
                }

            }
        });
    }

}

`;

exports[`macros getter & setter: getter & setter 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options get text() {
        return this._text;
    }
    @options set text( value ) {
        this._text = value;
    }
    
    @options get date() {
        return this._date;
    }
    @options set date( value ) {
        this._date = value;
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            text: {
                get() {
                    return this._text;
                },

                set(value) {
                    this._text = value;
                }

            },
            date: {
                get() {
                    return this._date;
                },

                set(value) {
                    this._date = value;
                }

            }
        });
    }

}

`;

exports[`macros getter: getter 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options get startDate() {
        return new Date();
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            startDate: {
                get() {
                    return new Date();
                }

            }
        });
    }

}

`;

exports[`macros method: method 1`] = `

import { options } from '../babel.macro';

class FooComponent extends Template {
    @options text() {
        return this.ID;
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { configureOptions } from 'sham-ui';


class FooComponent extends Template {
    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(FooComponent.prototype, this, {
            text() {
                return this.ID;
            }

        });
    }

}

`;
